### PR TITLE
feat: add bundled health check for KServe LLMInferenceService (#29599)

### LINKED ISSUE #29599: Add bundled health assessment for KServe LLMInferenceService
# Summary

Add a bundled Lua health check for `serving.kserve.io/LLMInferenceService` so Argo CD can assess the readiness reported by the KServe controller without requiring each installation to supply its own health customization.

In local Argo CD `master` at `6951cbc21a596f75b0867ea0530de2d0478ebe79`, `resource_customizations/serving.kserve.io/` contains an `InferenceService` check but no `LLMInferenceService` check. The upstream [KServe customization directory](https://github.com/argoproj/argo-cd/tree/master/resource_customizations/serving.kserve.io/InferenceService) was also checked through the GitHub API on 2026-09-07. No matching bundled wildcard or Go health handler was found in the local checkout.

This is a missing bundled assessment, not a demonstrated workload outage. There are no live `LLMInferenceService` instances in the current local cluster, so no failing workload, incorrect Application health transition, or blocked sync has been reproduced.

# Motivation

Teams managing LLM inference services through GitOps would benefit from seeing controller-reported readiness in Argo CD and using that assessment when coordinating dependent resources. A shared bundled check would reduce the need to maintain equivalent installation-specific scripts.

`LLMInferenceService` is a separate resource kind from `InferenceService`. [Issue #25263](https://github.com/argoproj/argo-cd/issues/25263) concerns the existing `InferenceService` check and its `Stopped` condition; it does not request support for this kind. Searches across upstream issues and PRs for `LLMInferenceService`, `llmisvc`, and `"LLM inference"` returned zero matches on 2026-09-07, and an open-PR search for `kserve` also returned zero.

# Proposal

Introduce `resource_customizations/serving.kserve.io/LLMInferenceService/health.lua`, a corresponding `health_test.yaml`, and resource manifests under `testdata/`. Keep this proposal limited to this kind and its own reported status; leave existing `InferenceService` behavior, resource actions, and other LLM resource kinds outside the scope.

Use the controller's aggregate readiness condition as the starting point. The [KServe v0.18.0 v1alpha2 API definition](https://github.com/kserve/kserve/blob/v0.18.0/pkg/apis/serving/v1alpha2/llm_inference_service_types.go) exposes the `Ready` condition and embeds the common status type. Its [condition implementation](https://github.com/kserve/kserve/blob/v0.18.0/pkg/apis/serving/v1alpha2/llm_inference_service_lifecycle.go) aggregates `WorkloadsReady` and `RouterReady`, but excludes `PresetsCombined` from that aggregate. Explicitly account for configuration-combination failures before returning `Healthy`; confirm the full reconciliation and generation semantics before finalizing the mapping or claiming compatibility with additional versions.

Proposed behavior for discussion:

- Report `Healthy` when readiness is true and, where the controller reliably reports it, the observed generation is current.
- Report `Progressing` while status or readiness is missing, readiness is unknown, or reconciliation has not caught up with the current generation.
- Distinguish documented failures from normal startup or reconciliation when readiness is false; agree the `Degraded` versus `Progressing` mapping using controller code and observed status examples.
- Report `Suspended` for the current `serving.kserve.io/stop` annotation when its value is case-insensitively `true`. KServe's [stop helper](https://github.com/kserve/kserve/blob/v0.18.0/pkg/utils/utils.go) and [LLM workload controller](https://github.com/kserve/kserve/blob/v0.18.0/pkg/controller/v1alpha2/llmisvc/workload_single_node.go) establish this signal for this kind. Treat a stale `Stopped` reason after removal of that annotation as resuming, not continued suspension. This proposed presentation describes requested intent, not completed cleanup.
- Preserve useful controller reason/message information and handle absent optional fields safely. Avoid treating unrelated group conditions or absent unused workload components as readiness requirements.

Proposed validation includes healthy, initializing, unknown, failed, stopped, resuming, and stale-generation cases; `PresetsCombined=False` alongside stale `Ready=True`; missing status/conditions/readiness; absent reason/message fields; and condition-order independence. Capture representative full manifests with complete status from an environment running the controller, record its version, and distinguish any synthetic edge cases from captured output. These fixtures are not available yet in the current cluster. Run the existing Lua test suite and add Application-level e2e coverage for the resulting visible health transitions and any claimed sync-wave behavior; if such coverage is infeasible, explain the specific reason in the PR and discuss it with maintainers.

Update the target release's upgrade notes under `Custom Healthchecks Added` and document the supported status semantics and version assumptions. Follow the repository-required build, code generation, lint, test, and CLI checks before proposing a commit. This draft contains no implementation or completed validation claims.

This proposal was prepared with AI assistance and source inspection. It requests enhancement triage; maintainer acceptance of the health policy, human review, and fixture collection remain prerequisites for an implementation contribution.

### PR DESCRIPTION
**Pull request by sarfarazgit**
Closes https://github.com/argoproj/argo-cd/issues/29599

Add a bundled Lua health assessment for serving.kserve.io/LLMInferenceService so Argo CD can report readiness from the KServe controller without requiring each installation to supply its own health customization.

The LLMInferenceService status embeds Knative duckv1 conditions: the dependent conditions (PresetsCombined, WorkloadsReady, RouterReady) aggregate into the top-level Ready condition. Health is derived from Ready, with a stopped service (reason "Stopped") mapped to Suspended and not-ready dependents surfaced in the message. Includes testdata fixtures for healthy, progressing, degraded and stopped states.

What was added:

[health.lua]  the health assessment
[health_test.yaml] 5 test cases
5 testdata/ fixtures (healthy, progressing, progressing-with-no-status, degraded, stopped)
All 5 cases pass via the existing TestLuaHealthScript harness in [util/lua/health_test.go]

How the check works:

I verified the KServe upstream CRD model (via the GitHub API on the kserve/kserve master tree, since this is a newer CRD than the existing InferenceService). Key facts that shaped the logic:

- LLMInferenceServiceStatus embeds Knative duckv1.Status, so it's a "living" resource: dependent conditions (PresetsCombined, WorkloadsReady, RouterReady, defined in pkg/apis/serving/v1alpha2/llm_inference_service_lifecycle.go) aggregate into the top-level Ready condition.

- The controller signals a stopped service by setting condition reason: Stopped on its workload/preset conditions (confirmed in controller_int_stop_test.go), rather than a dedicated Stopped condition type like the older InferenceService uses.

So the assessment:

[ ] - No status / no conditions → Progressing ("waiting to report")
[ ] - Any condition with reason: Stopped → Suspended
[ ] - Ready = True → Healthy
[ ] - Ready = False → Degraded, surfacing the not-ready dependents in the message
[ ] - Ready = Unknown or absent → Progressing

Notes:

This is purely additive (a missing bundled assessment), and I could not reproduce a live workload against it, there are no LLMInferenceService instances in the local cluster, so the verification is the unit-test fixtures rather than an observed Application health transition.

I checked the docs: [docs/operator-manual/health.md](vscode-webview://1e8c8p6aj6d7m2nmkopl6p9dv01dh44g32kjle0msvibft1ahlte/docs/operator-manual/health.md) documents how bundled checks work but does not maintain a per-resource list (the existing InferenceService check isn't listed either), so no docs change is warranted.

### CODE REVIEW

**Review by sarfarazgit — COMMENTED** [review]
@anandf @ppapapetrou76 @nitishfy  - Could you please review this PR and share your feedback when you have a moment?

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua [hunk 1]
```diff
@@ -0,0 +1,66 @@
+local health_status = {}
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:1]: Fixed

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+health_status.message = "Waiting for LLMInferenceService to report status..."
+
+if obj.status == nil or obj.status.conditions == nil then
+  return health_status
+end
+
+-- KServe reconciles the LLMInferenceService as a Knative "living" resource: the
+-- dependent conditions (PresetsCombined, WorkloadsReady, RouterReady and their
+-- sub-conditions) are aggregated into the top-level "Ready" condition. We assess
+-- health from "Ready" and surface the not-ready dependents in the message.
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:13]: Stale? I don't see any vscode-webview link.

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua [hunk 3]
```diff
@@ -0,0 +1,66 @@
+local health_status = {}
+
+health_status.status = "Progressing"
+health_status.message = "Waiting for LLMInferenceService to report status..."
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:4]: It's fine.

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+if stopped then
+  health_status.status = "Suspended"
+  health_status.message = "LLMInferenceService is Stopped"
+  return health_status
+end
+
+if ready ~= nil and ready.status == "True" then
+  health_status.status = "Healthy"
+  health_status.message = "LLMInferenceService is healthy."
+  return health_status
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:75]: Not necessary for just adding a health check. (I updated the rule on the Qodo backend to exclude PRs that only add health checks.)

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+end
+
+if ready ~= nil and ready.status == "False" then
+  health_status.status = "Degraded"
+else
+  -- Ready is Unknown or has not been reported yet.
+  health_status.status = "Progressing"
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:82]: Not necessary for a health check. Updated the Qodo rule.

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+if stopped then
+  health_status.status = "Suspended"
+  health_status.message = "LLMInferenceService is Stopped"
+  return health_status
+end
+
+if ready ~= nil and ready.status == "True" then
+  health_status.status = "Healthy"
+  health_status.message = "LLMInferenceService is healthy."
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:74]: @sarfarazgit this is valid, can you add observedGeneration handling (assuming the controller implements observedGeneration, either at the status or conditions level)? There are examples in other health checks.

**Comment by crenshaw-dev** [comment]
@sarfarazgit can you fix the signoff/DCO check?

File: resource_customizations/serving.kserve.io/LLMInferenceService/health.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (sarfarazgit) [reply resource_customizations/serving.kserve.io/LLMInferenceService/health.lua:74]: @crenshaw-dev Sure! I will do the needful.

**Review by crenshaw-dev — APPROVED** [review]
Thanks!

**Comment by crenshaw-dev** [comment]
DCO is still failing... I've always found it easiest to `git reset HEAD~<N>` and then commit all my changes in one commit with `--signoff`. Then just force-push.

**Comment by sarfarazgit** [comment]
> DCO is still failing... I've always found it easiest to `git reset HEAD~<N>` and then commit all my changes in one commit with `--signoff`. Then just force-push.

It’s done. Thank you for your advice. it was very helpful and much appreciated.
